flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorSubtaskState.java

```diff
@@ -227,6 +229,14 @@ private static void registerSharedState(
         for (KeyedStateHandle stateHandle : stateHandles) {
             if (stateHandle != null) {
                 stateHandle.registerSharedStates(sharedStateRegistry, checkpointID);
+                // Registering state handle to the given sharedStateRegistry serves two purposes:
+                // 1. let sharedStateRegistry be responsible for cleaning the state handle,
```

Comment:
I think this can be avoided because checkpoint will be discarded once no state is in use, which will also discard the private state.
And it's better to keep discarding there (in `CheckpointCleaner`), rather than in `SharedStateRegistry` because the back-pressure mechanism is better there.

So I propose to have an empty `discardState` method in wrapper.
(That would also allow to replace `keyedStateHandle` in it with only an ID)

Reply:
The ID is formally necessary to implement `equals` (`hashCode`). Otherwise, each time this ID is used, `SharedStateRegistry` will try to discard it.
Discard is a no-op, but currently deletions are scheduled one-by-one to the IO thread pool. So it might delay other IO tasks and/or lengthen the queue inside the pool.

Reply:
You are right, I added the ID to the wrapper class.